lib/settings/settings.js

```diff
@@ -0,0 +1,85 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const assert  = require('assert')
+
+const KEY = Symbol()
+const POLL_INTERVAL = Symbol()
+
+module.exports = (config, mc, log) => {
+
+
+  // A sentinel error for signaling that the result was invalid/missing,
+  // and we should try to push our own representation to memcached.
+  class PushOnMissing extends Error {}
+
+  // An abstract class that stores options in memcached.
+  //
+  // Others extend this class to hotload options from memcached.
+  class Settings {
+
+    constructor(key) {
+      assert(typeof key === 'string')
+      this[KEY] = key
+    }
+
+    // subclasses should provide their own implementation
+    setAll(value) {
+      return this
+    }
+
+    // subclasses should provide their own implementation
+    validate(value) {
+      return value
+    }
+
+    // Pushes `this` as JSON to `key`
+    //
+    // Customize what is pushed with a toJSON method.
+    push() {
+      log.info({ op: this[KEY] + '.push' })
+      return mc.setAsync(this[KEY], this, 0)
+        .then(() => this.refresh())
+    }
+
+    refresh(options) {
+      log.info({ op: this[KEY] + '.refresh' })
+      let result = mc.getAsync(this[KEY]).then(value => this.validate(value))
+
+      if (options && options.pushOnMissing) {
+        result = result.catch(err => {
+          if (err instanceof PushOnMissing) {
+            log.info({ op: this[KEY] + '.refresh.pushOnMissing' })
+            return this.push()
+          } else {
+            throw err
+          }
+        })
+      }
+      return result.then(
+        value => this.setAll(value),
+        err => {
+          log.error({ op: this[KEY] + '.refresh', err: err })
+          throw err
+        }
+      )
+    }
+
+    pollForUpdates() {
+      this.stopPolling()
+      this[POLL_INTERVAL] = setInterval(this.refresh.bind(this), config.updatePollIntervalSeconds * 1000)
+      this[POLL_INTERVAL].unref()
+    }
+
+    stopPolling() {
+      clearInterval(this[POLL_INTERVAL])
```

Comment:
It's the JavaScript way of private properties. The nodejs source has been growing their use of `Symbol` for any new private properties they need (can't really remove all the `_prop`, even though the underscore was supposed stop people from using, but lol).

There was 2 reasons to use these here:

1. These properties should be private, there's no reason for a sub class to fiddling with them.
2. Properties set by a `Symbol` aren't enumerable, so they don't get serialized into JSON. This could be done with `Object.setProperty`, but `Symbol` seemed better.

Farther Future Us should hopefully be quite used to `Symbol` in nodejs code, but I can add a comment for Short Future Us if that seems valuable.